src/semantic/type_checker.rs

```diff
@@ -0,0 +1,161 @@
+#![allow(unused_variables)]
+use lexical::types::Types;
+use syntax::expr::ExprWrapper;
+use syntax::expr::Expr::*;
+use semantic::analyzer_trait::ASTAnalyzer;
+
+pub struct TypeChecker;
+
+impl TypeChecker {
+    pub fn new() -> TypeChecker {
+        TypeChecker
+    }
+
+    fn cmp_lhs_rhs(lhs_str: String, rhs_str: String) -> Option<String> {
+        match (lhs_str.parse::<Types>(), rhs_str.parse::<Types>()) {
+            (Ok(lhs_type), Ok(rhs_type)) => {
+                if lhs_type != rhs_type {
+                    panic!("Error goes here"); // FIXME: Better errors
+                }
+
+                Some(lhs_str)
+            },
+            (Err(()), Err(())) => {
+                // TODO: Custom type comparison. May or may not be an error.
+                panic!("Found a custom type references");
+            },
+            // Found a builtin type and a custom type -> error
+            _ => panic!("Error goes here") // FIXME: Better errors
+        }
+    }
+}
+
+impl ASTAnalyzer<Option<String>> for TypeChecker {
+    fn analyze(&mut self, ast_root: &mut ExprWrapper) -> Option<String> {
+        // TODO: Handle errors, Result<ExprWrapper, String>?
+        // How to get Expr's type?
+        // Basics:
+        // Literal -> Builtin Type (Done) or Custom Type
+        // Variable -> VarName -> Lookup VarDecl -> Type
+        // FnCall -> FnName -> LookUp FnDecl -> Type
+        // InfixOp -> analyze bubble up type from Expr op Expr
+        // UnaryOp -> analyze bubble up type from Expr
+
+        // Future?:
+        // if rust style let i = if a {} else {}
+        // If needs lookup all returns in all possible code blocks, assert they're the same type
+
+        match *ast_root.get_mut_expr() {
+            Assign(ref mut var_name_expr_wrapper, ref mut rhs_expr_wrapper) => {
+                let lhs_type = match self.analyze(var_name_expr_wrapper) {
```

Comment:
regardless of if it should never happen, if it *does* happen, you'll just panic there, and be like, oh okay, i was wrong this will happen. so if `unreachable!` also panics, then there's little to no difference between the two, even if you aren't sure